src/types.ts

```diff
@@ -0,0 +1,283 @@
+export const asStoreSettings = asObject({
+  ipWhitelist: asMap(asBoolean),
+  apiKeyWhitelist: asMap(asBoolean)
+})
+export const asStoreSettingsDocument = asObject({
+  ...nanoDocumentShape,
+  ...asStoreSettings.shape
+})
+
+// Directory
+
+export type StoreDirectory = ReturnType<typeof asStoreDirectory>
+export type StoreDirectoryDocument = ReturnType<typeof asStoreDirectoryDocument>
+export const asStoreDirectory = asObject({
+  timestamp: asTimestampRev,
+  deleted: asStoreFileTimestampMap,
+  paths: asStoreFileTimestampMap,
+  timestampHistory: asTimestampHistory
+})
+export const asStoreDirectoryDocument = asObject({
+  ...storeMergeDocumentShape,
+  ...asStoreDirectory.shape
+})
+
+// Repo
+
+export interface StoreRepo extends StoreDirectory {
+  timestamp: TimestampRev
+  lastGitHash?: string
+  lastGitTime?: number
+  size: number
+  sizeLastCreated: number
+  maxSize: number
+}
+export type StoreRepoDocument = ReturnType<typeof asStoreRepoDocument>
+export const asStoreRepo = asObject<StoreRepo>({
+  ...asStoreDirectory.shape,
+  lastGitHash: asOptional(asString),
+  lastGitTime: asOptional(asNumber),
+  size: asNumber,
+  sizeLastCreated: asNumber,
+  maxSize: asNumber
+})
+export const asStoreRepoDocument = asObject({
+  ...storeMergeDocumentShape,
+  ...asStoreRepo.shape
+})
+
+// File
+
+export type StoreFile = ReturnType<typeof asStoreFile>
+export type StoreFileDocument = ReturnType<typeof asStoreFileDocument>
+export const asStoreFile = asObject({
+  timestamp: asTimestampRev,
+  box: asEdgeBox
+})
+export const asStoreFileDocument = asObject({
+  ...storeMergeDocumentShape,
+  ...asStoreFile.shape
+})
+
+// Change Set
+
+export type FileChange = ReturnType<typeof asFileChange>
+export const asFileChange = asEither(asObject({ box: asEdgeBox }), asNull)
+
+export type ChangeSet = ReturnType<typeof asChangeSet>
+export const asChangeSet = asMap(asFileChange)
+
+// Union of all store data types
+export type StoreData = StoreSettings | StoreRepo | StoreDirectory | StoreFile
+// Union of all document types
+export type StoreDocument =
+  | StoreSettingsDocument
+  | StoreRepoDocument
+  | StoreDirectoryDocument
+  | StoreFileDocument
+
+// API Types:
+
+export type StoreFileWithTimestamp = ReturnType<typeof asStoreFileWithTimestamp>
+export const asStoreFileWithTimestamp = asObject({
+  ...asStoreFile.shape,
+  timestamp: asTimestampRev
+})
+
+export type StoreDirectoryPathWithTimestamp = ReturnType<
+  typeof asStoreDirectoryPathWithTimestamp
+>
+export const asStoreDirectoryPathWithTimestamp = asObject({
+  paths: asStoreFileTimestampMap,
+  timestamp: asTimestampRev
+})
+
+export type GetFilesMap = ReturnType<typeof asGetFilesMap>
+export const asGetFilesMap = asMap(
+  asEither(asStoreFileWithTimestamp, asStoreDirectoryPathWithTimestamp)
+)
+
+// API Responses
+
+export type ApiResponse<Data> = ApiOkResponse<Data> | ApiErrorResponse
+
+export interface ApiOkResponse<Data> {
```

Comment:
I would do:

```js
const asApiResponse = asObject({
  success: asBoolean,
  message: asOptional(asString, "Success"),
  error: asOptional(asString),
  data: asUnknown
})
```

Then you can clean the response right away, figure out success or failure, and then clean `data` as a second step. However, we can leave this for now, since I realize that would probably involve refactoring your other repo.

Reply:
So, data is typed; not unknown. Currently typescript generics is being used to define the type. Since what I have been calling "higher-order cleaners" seem homomorphic to what typescript calls a "generic", I wonder if there is a way to define `asApiResponse` similar to how `ApiResponse<Data>` is defined.

Reply:
Well, you certainly knocked it out of the park.